### PR TITLE
Group chat

### PR DESCRIPTION
**Pull request by DanielC34**
'/groups' endpoint fetching in addUserModel is bringing up error 404

### CODE REVIEW

File: nextjs/app/chats/components/models/addGroupModel.js [hunk 1]
```diff
@@ -43,14 +43,14 @@ const AddGroupUserModel = ({ onClose }) => {
           console.log(`Creating group "${groupName}" with users:`, selectedUsers);
 
           //Make API call to create the group
-          const response = await fetch("/groups", {
+          const response = await fetch("/messages/groups", {
```
Comment (brightonc12) [inline nextjs/app/chats/components/models/addGroupModel.js:46]: @DanielC34  why do you need to fetch groups?

File: server/routes/messages.js [hunk 2]
```diff
@@ -6,6 +6,33 @@ const mustBeAuthenticated = require("../middleware");
 //Middleware to ensure authentication
 messagesRouter.use(mustBeAuthenticated);
 
+//Route to create a new group (POST /groups)
+messagesRouter.post('/groups', async (req, res) => {
+    try {
+      //Extract necessary data from request body
+      const { groupName, selectedUsers } = req.body;
```
Comment (brightonc12) [inline server/routes/messages.js:13]: `selectedUsers` should be `user_ids`

File: server/routes/messages.js [hunk 3]
```diff
@@ -6,6 +6,33 @@ const mustBeAuthenticated = require("../middleware");
...
+
+      if (!groupName || !selectedUsers || Array.isArray(selectedUsers)) {
+        return res
+          .status(400)
+          .send({ msg: "Invalid request. Missing required data" });
+      }
+
+      //Create a new group
+      const group = new Chat({ name: groupName, members: selectedUsers });
```
Comment (brightonc12) [inline server/routes/messages.js:22]: change name from `group` to `chat`

File: server/routes/messages.js [hunk 4]
```diff
@@ -6,6 +6,33 @@ const mustBeAuthenticated = require("../middleware");
...
+
+      //Save the group in the database
+      await group.save();
```
Comment (brightonc12) [inline server/routes/messages.js:25]: change name from group to chat

File: server/routes/messages.js (hunk 3 above)
Comment (brightonc12) [reply server/routes/messages.js:22]: @DanielC34, where are you getting the `members` and `name` arguments from? chat schema doesn't have such fields

Comment (brightonc12) [reply server/routes/messages.js:22]: Chat schema only has `type` and `participants`

Comment (DanielC34) [reply server/routes/messages.js:22]: So what should line 22 contain? like inside the curly braces of new Chat

Comment (brightonc12) [reply server/routes/messages.js:22]:
```suggestion
      const group = new Chat({ type:"group", participants: selectedUsers, name: groupName });
```
@DanielC34  You will need to introduce a new field called name with the schema. the name will store the group name for group change and will be black for private chats

Comment (DanielC34) [reply server/routes/messages.js:22]: Wait did you not say i should change selectedUsers to user._id earlier?

Comment (brightonc12) [reply server/routes/messages.js:22]: No, i meant the variable name should change to `userIds` or `participantIds` or `recipientIds`

Comment (DanielC34) [reply server/routes/messages.js:22]: so for name in the chat model file, name should be a required string right? am i missing anything on that part?

Comment (brightonc12) [reply server/routes/messages.js:22]: When you create the name field, is should not be required on the schema level,

Comment (DanielC34) [reply server/routes/messages.js:22]: so it should be like this: name: { type: String },

Comment (brightonc12) [reply server/routes/messages.js:22]: yes

Comment (DanielC34) [reply server/routes/messages.js:22]: Now after making this change, where do i look into next relating to this? I just changed all the variable names and did as you said

Comment (DanielC34) [reply server/routes/messages.js:22]: Weird how when i try to load up the add group card its not bringing up the users after i made these changes

File: server/models/chat.js [hunk 5]
```diff
@@ -1,10 +1,12 @@
 const { Schema, model } = require("mongoose");
 
+
 // ChatSchema defined using mongoose Schema
 const chatSchema = new Schema(
     {
         type: {type: String, required: true}, //Type field is defined as required string
-        participants: [{type: Schema.Types.ObjectId, ref: "User", required: true}] //Participants defined as required array of User IDs for referencing in validation
+        participants: [{type: Schema.Types.ObjectId, ref: "User", required: true}], //Participants defined as required array of User IDs for referencing in validation
+        name: { type: String, required: true },
```
Comment (brightonc12) [inline server/models/chat.js:9]: `name`  should not be required cause private chats don't have a name

File: server/routes/messages.js [hunk 6]
```diff
@@ -1,11 +1,47 @@
...
+          .send({ msg: "Invalid request. Missing required data" });
+      }
+        
+    const users = await User.find({ _id: { $in: userIds } });
+
+      //Create a new group
+      const chat = new Chat({ type: "group", participants: userIds, name: groupName });
+
+      //Save the group in the database
+        await chat.save();
```
Comment (brightonc12) [inline server/routes/messages.js:28]: Check out https://github.com/DanielC34/convocafe/blob/main/server/routes/messages.js#L94. 
that's how you populate users

File: server/routes/messages.js [hunk 7]
```diff
@@ -91,7 +127,7 @@ messagesRouter.get('/chats', async (req, res) => {
 
     try {
         //Fetch chats where the user is a participant and populate participants' usernames
-        const chats = await Chat.find({participants: userId}).populate('participants', 'username');
+        const chats = await Chat.find({participants: userId}).populate('participants', 'username email');
```
Comment (brightonc12) [inline server/routes/messages.js:130]: username and email shouldn't be in a single string

**Pull request merged**